test/capi/capi_models.cxx

```diff
@@ -280,6 +279,220 @@ BOOST_AUTO_TEST_CASE(test_boosted_trees_double) {
+      CAPI_CHECK_ERROR(error);
+
+
+      tc_release(fd);
+    }
+
+    {
+      tc_flex_dict* fd = tc_flex_dict_create(&error); 
+      CAPI_CHECK_ERROR(error);
+
+      tc_parameters_add_flex_dict(args, "extra_data", fd, &error);
+      CAPI_CHECK_ERROR(error);
+
+
+      tc_release(fd);
+    }
+
+    tc_model* model; 
+
+      // We now have enough to create the model.
+      model = tc_model_new(model_name, &error);
+      CAPI_CHECK_ERROR(error);
+
+      tc_variant* ret = tc_model_call_method(model, "train", args, &error);
+      CAPI_CHECK_ERROR(error);
+      TS_ASSERT(ret != NULL);
+
+      tc_release(ret);
+    
+
+    tc_release(args);
+
+    TS_ASSERT(model != NULL);
+
+    std::string ret_name = tc_model_name(model, &error);
+
+    TS_ASSERT(ret_name == model_name);
+    
+
+    // Test predictions on the same data.  Should be almost completely
+    // accurate...
+    {
+      std::vector<std::pair<std::string, std::vector<int64_t> > > rec_data = {
+        {"user_id", {1}}};
+
+      tc_sframe* sf = make_sframe_integer(rec_data);
+
+      CAPI_CHECK_ERROR(error);
+
+      tc_parameters* p_args = tc_parameters_create_empty(&error);
+
+      tc_parameters_add_sframe(p_args, "query", sf, &error);
+      CAPI_CHECK_ERROR(error);
+
+      tc_parameters_add_int64(p_args, "top_k", 2, &error);
+      CAPI_CHECK_ERROR(error);
+
+      {
+        tc_variant* ret_2 = tc_model_call_method(model, "recommend", p_args, &error);
+        CAPI_CHECK_ERROR(error);
+
+        tc_release(p_args);
+
+        int is_sframe = tc_variant_is_sframe(ret_2);
+        TS_ASSERT(is_sframe);
+
+        tc_sframe* sa = tc_variant_sframe(ret_2, &error);
+        CAPI_CHECK_ERROR(error);
+        size_t n_rows = tc_sframe_num_rows(sa, &error);
+
+        TS_ASSERT(n_rows == 2);
+        CAPI_CHECK_ERROR(error);
+
+        tc_release(ret_2);
+      }
+
+      if(std::string(model_name) == "item_similarity") {
+        tc_parameters* export_args = tc_parameters_create_empty(&error);
+        CAPI_CHECK_ERROR(error);
+
+        {
+          std::string url = turi::fs_util::system_temp_directory_unique_path(
+            "", "_coreml_export_test_1_tmp.mlmodel");
+          tc_flexible_type* ft_name = tc_ft_create_from_cstring(url.c_str(), &error);
+          CAPI_CHECK_ERROR(error);
+          tc_parameters_add_flexible_type(export_args, "filename", ft_name,
+                                          &error);
+          CAPI_CHECK_ERROR(error);
+          tc_release(ft_name);
+        }
+
+        tc_model_call_method(model, "export_to_coreml",
+                             export_args, &error);
+        CAPI_CHECK_ERROR(error);
+        tc_release(export_args);
+      }
+    }
+
+    // Test saving and loading the model.
+    {
+      std::string model_path;
+      std::string error_message;
+      std::string expected_substr;
+
+#ifdef MECHANISM_FOR_TRIGGERING_AN_ERROR_IN_DOCKER
+      // sad path 1 - attempting to save without permission to location
```

Comment:
You don't need to test all the sad paths here - I think the purpose of that was actually to make sure our low level filesystem handling code had the correct behavior (mostly correct error messages) for various permission issues. Since this test is really testing the same thing across all different models, I would skip it for this one and go straight to the "happy path" (the one where the model actually saves), if you want to test save and load for this model.

Reply:
Thanks for letting me know! Sad paths removed :)